it/src/test/scala/quasar/datasource/kafka/KafkaDatasourceITSpec.scala

```diff
@@ -0,0 +1,178 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.datasource.kafka
+
+import slamdata.Predef._
+
+import java.util.UUID
+
+import org.apache.kafka.clients.producer.ProducerRecord
+import org.apache.kafka.common.serialization.{Serializer, StringSerializer}
+import org.specs2.mutable.Specification
+import org.specs2.specification.BeforeAfterAll
+import org.typelevel.jawn.AsyncParser
+
+import argonaut.Argonaut._
+import argonaut.JawnParser.facade
+import argonaut._
+import cats.data.EitherT
+import cats.effect.{IO, Resource}
+import cats.kernel.instances.uuid._
+import jawnfs2._
+import net.manub.embeddedkafka.{EmbeddedK, EmbeddedKafka, EmbeddedKafkaConfig}
+import quasar.api.datasource.DatasourceError.InitializationError
+import quasar.api.resource.{ResourceName, ResourcePath}
+import quasar.connector.datasource.LightweightDatasourceModule.DS
+import quasar.connector.{ByteStore, DataFormat, QueryResult}
+import quasar.qscript.InterpretedRead
+import quasar.{NoopRateLimitUpdater, RateLimiter, ScalarStages}
+
+class KafkaDatasourceITSpec extends Specification with BeforeAfterAll {
+  sequential
+
+  import KafkaDatasourceITSpec._
+
+  // Do not declare val's depending on this
+  var embeddedK: EmbeddedK = _
+
+  override def beforeAll(): Unit = {
+    embeddedK = EmbeddedKafka.start()(EmbeddedKafkaConfig(kafkaPort = 0, zooKeeperPort = 0))
+    implicit val config: EmbeddedKafkaConfig = embeddedK.config
+    implicit val stringSerializer: Serializer[String] = new StringSerializer
+
+    EmbeddedKafka.createCustomTopic("valueOnly")
+    EmbeddedKafka.createCustomTopic("keyAndValue")
+    EmbeddedKafka.createCustomTopic("partitioned", partitions = 5)
+
+    EmbeddedKafka.withProducer[String, String, Unit] { producer =>
+      producer.send(new ProducerRecord("valueOnly", s"{ ${q("key")}: ${q("value")} }"))
+      producer.send(new ProducerRecord("valueOnly", "[1, 2, 3]"))
+      producer.send(new ProducerRecord("valueOnly", q("string")))
+
+      producer.send(new ProducerRecord("keyAndValue", q("key"), q("value")))
+      producer.send(new ProducerRecord("keyAndValue", "[1, 2, 3]", "true"))
+
+      for (number <- 1 to 50) producer.send(new ProducerRecord("partitioned", s"""{ "number": $number }"""))
+    }
+  }
+
+  "Datasource" >> {
+    def baseConfig = Json(
+      "bootstrapServers" := List(s"localhost:${embeddedK.config.kafkaPort}"),
+      "groupId" := "precog",
+      "format" := Json(
+        "type" := "json",
+        "variant" := "line-delimited",
+        "precise" := false)
+    )
+
+    "reads value only topics" >> {
```

Comment:
Could you add an additional test for the case when the decoder is `Decoder.rawKey.asJson` but there are no keys in the data:
```scala
      def config =
        ("topics" := List("valueOnly", "partitioned")) ->:
        ("decoder" := Decoder.rawKey.asJson) ->:
        baseConfig
```
Would that return no output data?

Reply:
It would throw an NPE, most likely. In that case we get a `null`.

Reply:
I mean, the _test_ would throw an NPE, because it tries to parse it. The connector would just pass the `null` to quasar.

Reply:
Does the connector pass a JSON `null` or a JVM `null` to Quasar?

Quasar isn't equipped to recover gracefully from a JVM `null`, and would throw a runtime exception, which we should avoid. Additionally, we want to provide a helpful error to the user in this case, as this situation could imply they misconfigured the datasource. This means we'd handle this as an error during evaluation in the datasource and would not pass JVM `null` to Quasar.

A related question: is it possible for a topic to contain some entries which include both a key and a value, and some entries which only include a value? In this situation, if the user selects `Decoder.rawKey`, what would happen now? If this is a situation that can happen in Kafka, I think we'd want to ignore the entries without a key, and just return the keys that exist.

Reply:
@dcsobral and I resolved this: we are omitting the entries without a key. (We were already doing this when I wrote this last comment as well, I was just confused :) )

Reply:
It won't pass `null` to Quasar. It will treat `null` as empty. This is a decoder decision, so a different decoder can choose to handle it differently.